DOCS-2881: Revamp Micro-RDK docs (#4241)
viamrobotics/docs: docs/operate/get-started/setup-micro.md

Author: Note for reviewers: This page was moved to _index.md because it needed to become a folder in order for troubleshooting to be moved. If only GH could tell the true diff ;(

If you would like to instead retain the setup script, replace `/dev/null` in the above command with the location where you would like the script to be written, or remove the `-f /dev/null` entirely and the file will be written to `$HOME/export-esp.sh` by default.
{{% /hiddencontent %}}

### Build your firmware

Create firmware that integrates an existing module with the Micro-RDK:

1. Create a new machine and obtain its credentials:

   Navigate to the [Viam app](https://app.viam.com) and add a new machine.
   Click on the name of the machine to go to the machine's page, then select the **CONFIGURE** tab.

   Then select the part status dropdown to the right of your machine's name on the top of the page and copy the **Machine cloud credentials**:

   {{<imgproc src="/get-started/micro-credentials.png" resize="450x" declaredimensions=true alt="Machine part info menu accessed by Live status indicator, with machine cloud credentials button highlighted." class="shadow" >}}

   The Micro-RDK needs these credentials, which contains your machine part secret key and cloud app address, to connect to the [Viam app](https://app.viam.com).

>    {{% snippet "secret-share.md" %}}

Author: So, they don't have to enter it in--can still build without, and will pull in from env vars.

>    {{% snippet "secret-share.md" %}}

Reviewer: I see... would still leave this out. We might remove the env variable flow in this upcoming project: https://viam.atlassian.net/browse/RSDK-10503

1. Generate a new project skeleton from [this template](https://github.com/viamrobotics/micro-rdk/tree/main/templates/project):

   ```sh { class="command-line" data-prompt="$"}
   cargo generate --git https://github.com/viamrobotics/micro-rdk.git
   ```

   Follow the prompts:

>    - Select `templates/project` when prompted.
>    - Give the project a name of your choice.
>    - Select `esp32` for **MCU**.
>    - If you wish to configure an `esp32-camera` or a `fake` camera as a component of your machine, select **true** for **Include camera module?**.
>    - For **Machine cloud credentials**, paste in the machine cloud credentials you obtained in step 1.

Author: Ah good point I agree
```suggestion
   - Which sub-template should be expanded?:  `templates/project`.
   - Project name: Your choice.
   - MCU: `esp32`.
   - Include camera module?: If you will use an `esp32-camera` or a `fake` camera as a component of your machine, select `true`.
   - Machine cloud credentials: Paste in the machine cloud credentials you obtained in step 1
   - Enter WiFi name: The name of the WiFi network for the ESP32 to connect to on boot. Note that you must provide a 2.4 GHz network.
   - Enter the WiFi password: The password for the 2.GHz WiFi network.
```

   ```toml {class="line-numbers linkable-line-numbers" data-line="3"}
   [dependencies]
   ...
   micro-rdk-modular-driver-example = { git = "https://github.com/viamrobotics/micro-rdk.git", rev = "v0.5.0", package = "micro-rdk-modular-driver-example", features = ["esp32"] }
   ```

   {{% /tab %}}
   {{< /tabs >}}

1. Compile the project:

   ```sh { class="command-line" data-prompt="$"}
   make build-esp32-bin
   ```

   The first build may be fairly time consuming, as ESP-IDF must be cloned and built, and all dependent Rust crates must be fetched and built.
   Subsequent builds will be faster.

### Flash your ESP32

Upload the generated firmware to your ESP32:

1. Use a data cable to connect your ESP32 board to a USB port on your development machine.

1. Run the following command to flash the firmware to your ESP32:

   ```sh { class="command-line" data-prompt="$"}
   make flash-esp32-bin
   ```

   When prompted, select the serial port that your ESP32 is connected to through a data cable.

   If the flash is successful, you will retain a serial connection to the board until you press `Ctrl-C`.

>    To manage this connection, consider running it within a dedicated terminal session, or under `tmux` or `screen`.

Author: ```suggestion
```

   While the serial connection is live, you can also restart the currently flashed image with `Ctrl-R`.

1. Navigate to your new machine's page on the [Viam app](https://app.viam.com).
   If successful, the status indicator should turn green and show **Live**.

> After adding (or removing) a module or changing the version of a module, you must rerun the build and flash steps to rebuild the firmware and reflash the device.

Author: ```suggestion
```

### Configure and test your machine

You can now configure the models you included in your firmware and test them:

1. Navigate to your machine's page in the [Viam app](https://app.viam.com).

1. From the **CONFIGURE** tab, click **JSON** mode.
   Micro-RDK components and services must be configured in JSON.

1. Add your components and services to the machine configuration.
   If you are using one of the example modules, find configuration details in the [example modules README](https://github.com/viamrobotics/micro-rdk/blob/main/examples/modular-drivers/README.md#example-modules).
   For example, if you want to use a free heap sensor, your configuration could look like this:

   ```json
   {
     "components": [
       {
         "name": "my-free-heap-sensor",
         "api": "rdk:component:sensor",
         "model": "free-heap",
         "attributes": {}
       }
     ]
   }
   ```

1. Click **Builder** mode and find the configuration card of the component you want to test.

> 1. Click to open the **Test** section of the card and use the interface to test the component.

Author: ```suggestion
1. Click to open the **Test** section of the card and use the interface to test the component.

1. If you need to make changes to the module, make your changes, then rerun the build and flash steps to update the firmware and test the new version on your ESP32.
```

## Build firmware with cloud build

When developing and testing your firmware, it is easiest to build firmware locally.

> When you are ready to deploy your firmware to a fleet of microcontrollers, you can use the Viam cloud build service to simplify the process of building and hosting your firmware by eliminating the need to build locally and then upload the firmware image to a cloud storage bucket.

Reviewer: I'm noticing with cloud builds described above OTA, it's a little hard to understand why you would need to upload firmware to a cloud bucket.

I like how in the OTA section you have "Build OTA firmware locally." It might actually make sense to make this a "Build and host OTA firmware on GitHub" alternative.

Author: Yeah I felt the same way--I want to present them simultaneously, but they are distinct processes. Will have a think.

Author: Maybe we should actually *not* include Build OTA firmware locally, since it's an extra path most users don't need

Reviewer: I think it's okay to see how you would configure OTA and then how you would produce a URL - especially because I think we could also make the cloud build instructions shorter, like:

1. Create a GitHub repository for your firmware project. Note that any modules included in the firmware must point to GitHub repositories (not a local path).
2. [Current Step 5]
3. [Current Step 6]

In particular I don't think we should mention any of the credentials stuff (I appreciate you thinking about this though!) because the plan for https://viam.atlassian.net/browse/RSDK-10503 is to stop putting credentials in the firmware build, and instead only providing them at flash time via the micro-rdk-installer.

For including how to build OTA firmware locally - in a way I thought it was nice since it clarifies that OTA supports any URL, not just the one from cloud builds. And I liked the short and simple way you described it.

Author: Hmm I recognize that some of these steps are on the wordy side, but I think the explanation is helpful.

I think this proposed step 1 might be too abstract to be useful, at least for users who are still trying to understand how all the pieces fit together, e.g. modules vs firmware, and how cloud build works. Having just done this myself for the first time I would've found these steps useful.

Reviewer: Spoke offline - we'll simplify the steps and move the "cloud build" steps to below OTA as a "Build and host OTA firmware on GitHub"

Author: Made the changes to the order, and moved steps into tabs. Pared down the words in a few places, as well as removing the extra steps as discussed.

To set up cloud build, follow these steps:

1. Create a GitHub repository for any module you created and want to use in your firmware.
   Upload the contents of the module directory to the repository.

1. Update your firmware project's `Cargo.toml` file to use the URL of the GitHub repository, for example:

   ```toml
   [dependencies]
   ...

>    my-module = { git = "https://github.com/username/my-module.git" }

Reviewer: [note] I had wanted to use a concrete example module here, but I think it'll be better to do that after we move the example modules out of the micro-rdk repo, so this sounds fine for now

Feel free to resolve
